README
NAME
    IPC::PubSub - Interprocess Publish/Subscribe channels

SYNOPSIS
        # A new message bus with the DBM::Deep backend
        # (Other possible backends include Memcached and PlainHash)
        my $bus = IPC::PubSub->new(DBM_Deep => '/tmp/pubsub.db');

        # A channel is any arbitrary string
        my $channel = '#perl6';

        # Register a new publisher (you can publish to multiple channels)
        my $pub = $bus->new_publisher("#perl6", "#moose");

        # Publish a message (may be a complex object) to those channels
        $pub->msg("This is a message");

        # Register a new subscriber (you can subscribe to multiple channels)
        my $sub = $bus->new_subscriber("#moose");

        # Publish an object to channels
        $pub->msg("This is another message");

        # Set all subsequent messages from this publisher to expire in 30 seconds
        $pub->expiry(30);
        $pub->msg("This message will go away in 30 seconds");

        # Simple get: Returns the messages sent since the previous get,
        # but only for the first channel.
        my @msgs = $sub->get;

        # Simple get, with an explicit channel key (must be among the ones
        # it initially subscribed to)
        my @moose_msgs = $sub->get("#moose");

        # Complex get: Returns a hash reference from channels to array
        # references of [timestamp, message].
        my $hash_ref = $sub->get_all;

        # Changing the list of channels we subscribe to
        $sub->subscribe('some-other-channel');
        $sub->unsubscribe('some-other-channel');

        # Changing the list of channels we publish to
        $pub->publish('some-other-channel');
        $pub->unpublish('some-other-channel');

        # Listing and checking if we are in a channel
        my @sub_channels = $sub->channels;
        my @pub_channels = $pub->channels;
        print "Sub is in #moose" if $sub->channels->{'#moose'};
        print "Pub is in #moose" if $pub->channels->{'#moose'};

        # Raw cache manipulation APIs (not advised; use ->modify instead)
        $bus->lock('channel');
        $bus->unlock('channel');
        my @timed_msgs = $bus->fetch('key1', 'key2', 'key3');
        $bus->store('key', 'value', time, 30);

        # Atomic updating of cache content; $_ is stored back on the
        # end of the callback.
        my $rv = $bus->modify('key' => sub { delete $_->{foo} });

        # Shorthand for $bus->modify('key' => sub { $_ = 'val' });
        $bus->modify('key' => 'val');

        # Shorthand for $bus->modify('key' => sub { $_ });
        $bus->modify('key');

DESCRIPTION
    This module provides a simple API for publishing messages to *channels*
    and for subscribing to them.

    When a *message* is published on a channel, all subscribers currently in
    that channel will get it on their next "get" or "get_all" call.

    Currently, it offers three backends: "DBM_Deep" for on-disk storage,
    "Memcached" for possibly multi-host storage, and "PlainHash" for
    single-process storage.

    Please see the tests in t/ for this distribution, as well as "SYNOPSIS"
    above, for some usage examples; detailed documentation is not yet
    available.

SEE ALSO
    IPC::DirQueue, where the subscribers divide the published messages among
    themselves, so different subscribers never see the same message.

AUTHORS
    Audrey Tang <cpan@audreyt.org>

COPYRIGHT
    Copyright 2006 by Audrey Tang <cpan@audreyt.org>.

    This software is released under the MIT license cited below.

  The "MIT" License
    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



Changes
[Changes for 0.28 - 2008-08-15]

* We now require DBM::Deep 1.00 or later.

* Improved data consistency in DBM::Deep backend so it won't sometimes
  die with a "not an ARRAY reference" message.
  Reported by: Matthew Pitts

[Changes for 0.27 - 2007-10-09]

* Adjust tests so it no longer fails with newer versions of DBM::Deep.

[Changes for 0.26 - 2007-05-31]

* Repair a broken attempt at PAUSE uploading.

[Changes for 0.25 - 2007-05-31]

* The DBM::Deep tests of t/basic.t now explicitly passes the temporary
  database file to the backend; now it won't hangs on OSX anymore.
  Contributed by: Arne Skjærholt

[Changes for 0.24 - 2007-02-26]

* Typo fix in Memcached driver.

[Changes for 0.23 - 2007-02-19]

* New ->disconnect API to explicitly disconnect from the backend store.
  Contributed by: Alex Vandiver

[Changes for 0.22 - 2006-12-11]

* INCOMPATIBLE CHANGE to Jifty::DBI backend: "key" column changed to "data_key"
  because "key" is a reserved word in some SQL databases.
  Contributed by: Jesse Vincent

[Changes for 0.21 - 2006-10-26]

* Speed up ->modify calls for the Memcached backend.
* Normalized internal keys for channels, messages and data so they can't clash.

[Changes for 0.20 - 2006-10-25]

* Memcached: Remove the debug messages accidentally left in ->lock and ->unlock.
* New ->modify API for IPC::PubSub and Cache to atomically manipulate cache.
* Time::HiRes is now required to reduce locking contention.

[Changes for 0.11 - 2006-10-25]

* The Memcached backend now takes a namespace parameter to avoid collision.
* The ->lock, ->unlock, ->fetch, ->store APIs in IPC::PubSub now works again.

[Changes for 0.10 - 2006-10-25]

* Renamed from MessageBus.pm to IPC::PubSub.
* IPC::PubSub's factory methods are now ->new_subscriber and ->new_publisher.
* New ->channels API for Publisher and Subscriber objects.
* New ->publish, ->unpublish and ->expiry APIs for Publisher objects.
* New ->subscribe, ->unsubscribe APIs for Subscriber objects.

[Changes for 0.04 - 2006-10-24]

* Expose ->lock, ->unlock, ->fetch, ->store APIs into the IPC::PubSub object.
* Implement ->lock and ->unlock methods for non-Memcached backends.
* The tests are no longer entirely skipped when memcached is not running.

[Changes for 0.03 - 2006-10-24]

* New backend: JiftyDBI.
* Multiple publishers now work in DBM_Deep and Memcached backends.
* Memcached now atomically handles publisher announcement and removal.

[Changes for 0.02 - 2006-10-24]

* Thanks to mstrout++ this thing actually works now. :-)
* Switched from Class::InsideOut to Class::Accessor::Fast.

[Changes for 0.01 - 2006-10-24]

* Initial release to CPAN of this three-hours-old hack.


lib/IPC/PubSub.pm
package IPC::PubSub;
$IPC::PubSub::VERSION = '0.28';

use 5.006;
use strict;
use warnings;
use IPC::PubSub::Cacheable;
use IPC::PubSub::Publisher;
use IPC::PubSub::Subscriber;
use base qw/Class::Accessor::Fast/;

__PACKAGE__->mk_accessors(qw/_cache/);

sub new {
    my $self = bless {}, shift;

    my $backend = shift || 'PlainHash';

    local $@;
    eval { require "IPC/PubSub/Cache/$backend.pm" }
        or die "Cannot load backend module: IPC::PubSub::Cache::$backend: $@";

    $self->_cache(IPC::PubSub::Cacheable->new($backend => \@_));
    return $self;
}

sub new_publisher {
    my $self = shift;
    IPC::PubSub::Publisher->new($self->_cache, @_ ? @_ : '');
}

sub new_subscriber {
    my $self = shift;
    IPC::PubSub::Subscriber->new($self->_cache, @_ ? @_ : '');
}

sub fetch      { ( +shift )->_cache->fetch(@_) }
sub store      { ( +shift )->_cache->store(@_) }
sub lock       { ( +shift )->_cache->lock(@_) }
sub unlock     { ( +shift )->_cache->unlock(@_) }
sub modify     { ( +shift )->_cache->modify(@_) }
sub disconnect { ( +shift )->_cache->disconnect }

1;

__END__

=head1 NAME

IPC::PubSub - Interprocess Publish/Subscribe channels

=head1 SYNOPSIS

    # A new message bus with the DBM::Deep backend
    # (Other possible backends include Memcached and PlainHash)
    my $bus = IPC::PubSub->new(DBM_Deep => '/tmp/pubsub.db');

    # A channel is any arbitrary string
    my $channel = '#perl6';

    # Register a new publisher (you can publish to multiple channels)
    my $pub = $bus->new_publisher("#perl6", "#moose");

    # Publish a message (may be a complex object) to those channels
    $pub->msg("This is a message");

    # Register a new subscriber (you can subscribe to multiple channels)
    my $sub = $bus->new_subscriber("#moose");

    # Publish an object to channels
    $pub->msg("This is another message");

    # Set all subsequent messages from this publisher to expire in 30 seconds
    $pub->expiry(30);
    $pub->msg("This message will go away in 30 seconds");

    # Simple get: Returns the messages sent since the previous get,
    # but only for the first channel.
    my @msgs = $sub->get;

    # Simple get, with an explicit channel key (must be among the ones
    # it initially subscribed to)
    my @moose_msgs = $sub->get("#moose");

    # Complex get: Returns a hash reference from channels to array
    # references of [timestamp, message].
    my $hash_ref = $sub->get_all;

    # Changing the list of channels we subscribe to
    $sub->subscribe('some-other-channel');
    $sub->unsubscribe('some-other-channel');

    # Changing the list of channels we publish to
    $pub->publish('some-other-channel');
    $pub->unpublish('some-other-channel');

    # Listing and checking if we are in a channel
    my @sub_channels = $sub->channels;
    my @pub_channels = $pub->channels;
    print "Sub is in #moose" if $sub->channels->{'#moose'};
    print "Pub is in #moose" if $pub->channels->{'#moose'};

    # Raw cache manipulation APIs (not advised; use ->modify instead)
    $bus->lock('channel');
    $bus->unlock('channel');
    my @timed_msgs = $bus->fetch('key1', 'key2', 'key3');
    $bus->store('key', 'value', time, 30);

    # Atomic updating of cache content; $_ is stored back on the
    # end of the callback.
    my $rv = $bus->modify('key' => sub { delete $_->{foo} });

    # Shorthand for $bus->modify('key' => sub { $_ = 'val' });
    $bus->modify('key' => 'val');

    # Shorthand for $bus->modify('key' => sub { $_ });
    $bus->modify('key');

    # Disconnect the backend connection explicitly
    $bus->disconnect;

=head1 DESCRIPTION

This module provides a simple API for publishing messages to I<channels>
and for subscribing to them.

When a I<message> is published on a channel, all subscribers currently in
that channel will get it on their next C<get> or C<get_all> call.

Currently, it offers three backends: C<DBM_Deep> for on-disk storage,
C<Memcached> for possibly multi-host storage, and C<PlainHash> for
single-process storage.

Please see the tests in F<t/> for this distribution, as well as L</SYNOPSIS>
above, for some usage examples; detailed documentation is not yet available.

=head1 SEE ALSO

L<IPC::DirQueue>, where the subscribers divide the published messages among
themselves, so different subscribers never see the same message.

=head1 AUTHORS

Audrey Tang E<lt>cpan@audreyt.orgE<gt>

=head1 COPYRIGHT

Copyright 2006, 2007 by Audrey Tang E<lt>cpan@audreyt.orgE<gt>.

This software is released under the MIT license cited below.

=head2 The "MIT" License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.

=cut
